Handle world mods and world game mods in the config dialog.

lhofhansl opened the pull request:
See #10422

Minetest offers various per world options to locate mods:
1. `world/<worldid>/game/mods`
2. `world/<worldid>/worldmods`

In the both cases these mods are always selected, and in the first case no other mods can be selected.

With this change both cases are handled. For (1) all mods are displayed as game-content, and no other mods are shown. For (2) the mods are shown inline with all other mods installed and also displayed as game-content (i.e. those mods cannot be disabled)

This is a bugfix handling existing MT features correctly in the UI.

lhofhansl commented:
Not sure it's a feature. World game mods and worldmods already exist as features, but the UI does not handle them at all. It's more of a bug fix.

lhofhansl commented:
This is not quite right.
It handles world game mods (those placed in .../worlds/\<worldid>/game/mods) right.
But worldmods (.../worlds/\<worldid>/worldmods) is still not right. Should be its own category, but I do not understand FilterList enough to make it so, yet.

lhofhansl commented:
Ok... Please have a look now. Especially at the update sort function.
Worldmods are now sorted right before the Game mods.

lhofhansl commented:
This also fixes modpacks game packs - previously this would not indent correctly.

lhofhansl commented:
Comments? Concerns?

sfan5 commented on builtin/common/filterlist.lua:
```diff
@@ -290,8 +290,15 @@ function sort_mod_list(self)
 		if a.type ~= b.type or a.loc ~= b.loc then
```
There's a comment above here that should be updated to say `-- Show game and world mods at bottom`

lhofhansl replied:
Oops. Missed that. Thanks.

Also see #10422 where we discuss potentially removing the ".../game/mods" feature from MT.
I'd still like to get this change in. Can always remove it again.

Honestly the hardest part was go get the sorting right in this one, so that Game mods sort in the end and worldmods sort second to last.

lhofhansl replied:
#10439

lhofhansl commented:
This is what it looks like for world-mods (mods in the `../worlds/<world_id>/worldmods` folder).
"Worlsmods" are sorted second to last, and include modpacks are shown correctly.

![image](https://user-images.githubusercontent.com/446130/94635095-af9b5d00-0286-11eb-8ab4-339b932bbde3.png)

lhofhansl commented:
And here's with `../worlds/<world_id>/game/mods` folder. Only those mods are shown (as those are the only ones loaded when the game starts)

![image](https://user-images.githubusercontent.com/446130/94635348-4bc56400-0287-11eb-924f-36864a57c54d.png)

lhofhansl commented:
Updated to only handle worldmods correctly (but the game folder inside a world case)

lhofhansl commented:
Yes? No? :)

Let's handle at least worldmods correctly.

lhofhansl commented:
Fixed sorting again and added back world game mod handling - as long as MT supports mods like this, the UI should support it.
If we decide to remove world game mods (which I do not support) we can remove the UI handling of it as well.

Since there seems to be little interest in fixing bugs like this and also little interest allowing MT to ship self contained world experiences, I'll just say: **Take it or leave it**. :)

Not supporting this well is holding MT back. You can't ship exploration maps or premade arenas unless you save them as schematics yourself and add a mod to load them, which is both limiting and not very user friendly.

Not doing any more work on this unless there's more feedback.

lhofhansl commented:
I messed up my fork, so I deleted and re-created it.
